README.md
# cfpuppetserver

## Description

The package does standard procedure of installing Puppet Server, Puppet DB, PostgreSQL,
r10k, librarian-puppet and making it work all togather. It also lives in peace with
[cfsystem][] and [cfnetwork][] packages.

### Environment configuration

The configurations expects you to provide [Hiera version 4 configuration](https://docs.puppetlabs.com/puppet/latest/reference/lookup_quick.html#hierayaml-version-4-in-a-nutshell)
in environments. Example can be taken from [codingfuture/puppe-test](https://github.com/codingfuture/puppet-test/blob/master/hiera.yaml).

**NOTE1: there is a known closed bug in puppet <=4.3.2 - please make sure that all Hiera hierarchy files exist and both empty YAML and JSON files include '{}' at least**

## VERY IMPORTANT!!!

Now, the modules uses [cfdb][] for High Availability support out-of-the-box. The consequences is
that setup process is quite tricky - we need some facts and resources to be populated into Puppet DB
while Puppet DB is malfunctioning until the stack is fully configured.
Most likely, you will see some errors during conversion process and both Puppet DB & Puppet Server
**stop functioning**.

In that case, you can continue re-provisioning previously compiled catalog until Puppet Server 
can continue compiling new catalogs with the following command:

```bash
/opt/puppetlabs/bin/puppet apply --catalog /opt/puppetlabs/puppet/cache/client_data/catalog/$(/bin/hostname --fqdn).json
/opt/puppetlabs/bin/puppet agent --test
```

## Upgrade to puppetserver >= 2.5.0

As there is incompatible change related to bootstrap.cfg, please use `cfpuppetserver` < v0.10 for puppetserver < 2.5.

Upgrade procedures:

* Update to `cfpuppetserver` >= v0.10
* Manually deploy to current Puppet servers: `puppet agent -t`
* Puppet Server will fail to restart in 180 seconds
* Upgrade `puppetserver`/`puppetdb`/`puppet-agent` packages to latest versions
* Manually start services:
    * /bin/systemctl stop cfpuppetdb.service cfpuppetserver.service
    * /bin/systemctl start cfpuppetdb.service cfpuppetserver.service
* Wait for services to startup monitoring `netstat -pletn`
* Try Puppet deployment


## Global Hiera config

Puppet 4 has own implementation of lookup() which goes through:

* Global Hiera
* Per-environment Data Providers (Hiera, custom function)
* Per-module Data Providers (Hiera, custom function)

You should not use global Hiera any more. All configurations should be set in environments as mentioned above.

Global Hiera config is as follows:

```yaml
---
:backends:
  - yaml
:hierarchy:
  - global
:yaml:
  # Make sure to use hiera.yaml in environments
  :datadir: "/etc/puppetlabs/code/hieradata"
```

### Adding new Puppet clients

This module also provides a handy tool to initalize new puppet client hosts:

```
~#  /opt/codingfuture/bin/cf_gen_puppet_client_init
Usage: cf_gen_puppet_client_init <certname> [<cflocation> [<cflocationpool> [<http_proxy>]]
```

### Manual (re-)deployment of Puppet environments

```
~# /opt/codingfuture/bin/cf_r10k_deploy
```

### Automatic deployment via VCS (git) hook
```
~# ssh deploypuppet@puppet.example.com sudo /opt/codingfuture/bin/cf_r10k_deploy
```

### Manual fast refresh of PuppetServer JRuby instances

```
~# /opt/codingfuture/bin/cf_puppetserver_reload
```

### Manual purge of node trails from infrastructure

```
~# /opt/codingfuture/bin/cf_purge_node
```

### Restoring default node deactivation and purging behavior

Auto-configuration of CF modules heavily depends PuppetDB info. To protect from accident disasters,
automatic node info purging is disabled on purpose. It can be restored:

```
cfpuppetserver::puppetdb::settings_tune:
    database:
        'node-ttl': '7d'
        'node-purge-ttl': '14d'
```


## Setup

### Initial Puppet Server infrastructure

Either do manually (preferred for self-education) or use bundled [setup script](https://codingfuture.net/docs/cfpuppetserver/blob/master/setup_puppetserver.sh):
```
~# ./setup_puppetserver.sh
Usage: ./setup_puppetserver.sh <r10k_repo_url> [<certname=hostname> [<cflocation> [<cflocationpool> [<http_proxy=$http_proxy>] ] ] ]
```

### Config for Puppet Server node

Up to date installation instructions are available in Puppet Forge: https://forge.puppet.com/codingfuture/cfpuppetserver

Please use [librarian-puppet](https://rubygems.org/gems/librarian-puppet/) to deal with dependencies.
If this module is used for server setup then librarian-puppet is installed automatically.

There is a known r10k issue [RK-3](https://tickets.puppetlabs.com/browse/RK-3) which prevents
automatic dependencies of dependencies installation.

## Examples

Please check [codingufuture/puppet-test](https://github.com/codingfuture/puppet-test) for
example of a complete infrastructure configuration and Vagrant provisioning.


## `cfpuppetserver` class

* `deployuser = 'deploypuppet'` - user name for auto deploy user for VCS hook
* `deployuser_auth_keys = undef` - list of ssh_authorized_keys configurations
* `repo_url = undef` - repository location in URI format (e.g. ssh://user@host/repo or file:///some/path)
* `puppetserver = true` - if true then assume Puppet Server lives on this host (affects firewall)
* `puppetdb = true` - if true then assume Puppet DB lives on this host (affects firewall)
* `postgresql = true` - if true then PostgreSQL is setup on this node
* `autodiscovery = true` - if true then other instances of PuppetDB & PuppetServers are automatically
    discovered to be added for PuppetDB access / server_urls
* `puppetdb_hosts = []` - statically provide PuppetDB hosts (for autodiscovery = false)
* `puppetserver_hosts = []` - statically provide PuppetServer hosts (for autodiscovery = false)
* `client_hosts = []` - client hosts to allow access (dynamic ipset `cfpuppet_clients`)
* `iface = 'any'` - `cfnetwork::iface` name to listen for incoming client connections
* `cluster = 'cfpuppet'` - `cfdb` cluster to use
* `database = 'puppetdb' - `cfdb::database` to use in cluster
* `is_cluster = false` - goes directly to `cfdb::instance`
* `is_secondary = false` - goes directly to `cfdb::instance`
* `is_arbitrator = false` - goes directly to `cfdb::instance`
* `allow_update_check = false` - open firewall to connect to updates.puppetlabs.com, if enabled

## `cfpuppetserver::postgresql` class

NOTE: if PostgreSQL is setup through this module then you SHOULD NOT setup other cfdb instances
on the same node.

* `$settings_tune = {}`  - goes directly to `cfdb::instance`
* `$port = 5432` - goes directly to `cfdb::instance`
* `$node_id = undef` - required, if node ID cannot be retrieved from hostname in cluster mode
* `$password = undef` - force specific password instead of random generated
* `$version = undef` - version to pass to `cfdb::postgresql`
* `$memory_weight = 200` - goes directly to `cfdb::instance`
* `$memory_max = undef` - goes directly to `cfdb::instance`
* `$cpu_weight = 200` - goes directly to `cfdb::instance`
* `$io_weight = 200` - goes directly to `cfdb::instance`
* `$init_db_from = undef` - overrides default init_db_from, if set

## `cfpuppetserver::puppetdb` class

* `$use_proxy = 'secure'` - by default TLS channel is used for remote PostgreSQL connections. See `cfdb::access`.
* `$port = 8081` - port to use for PuppetDB instance
* `$max_connections = 30` - maximum number of connections per pool (there are two pools)
* `$memory_weight = 100` - relative weight for auto-distribution of memory resources
* `$memory_max = 256` - max memory in MB
* `$cpu_weight = 100` - relative weight for auto-distribution of CPU resources
* `$io_weight = 100` - relative weight for auto-distribution of I/O resources
* `$cert_whitelist = []` - specify the CNs of Puppet PKI to be accepted.
    If not set:
    * if Puppet Server runs the same node then `[$fqdn]`
    * otherwise, all nodes with Puppet Server configured
* `$settings_tune = {}` - a tree structure of PuppetDB INI for fine control

## `cfpuppetserver::puppetserver` class

* `$autosign = false` - DO NOT use in production. Enable auto-sign of client certificates.
* `$global_hiera_config = 'cfpuppetserver/hiera.yaml'` - default global Hiera config
* `$memory_weight = 100` - relative weight for auto-distribution of memory resources
* `$memory_max = undef` - max memory in MB
* `$cpu_weight = 100` - relative weight for auto-distribution of CPU resources
* `$io_weight = 100` - relative weight for auto-distribution of I/O resources
* `$activesupport_ver = '4.2.7.1'` - version of activesupport gem to install
* `$strict = 'warning'` - goes directly to puppet.conf
* `$disable_warnings = 'deprecations'` - goes directly to puppet.conf
* `$settings_tune = {}` - optional tune, tree structure
    * 'puppetserver'
        * 'max-active-instances' = $::facts['processorcount'] + 1
        * 'max-requests-per-instance' = 1000
        * 'compile-mode' = 'off'
        * 'connect-timeout-milliseconds' = 15000
        * 'idle-timeout-milliseconds' = 600000


[cfnetwork]: https://codingfuture.net/docs/cfnetwork
[cfsystem]: https://codingfuture.net/docs/cfsystem


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This
project adheres to [Semantic Versioning](http://semver.org/).

## 1.3.1 (2019-11-13)
- FIXED: to install ca-certificates in client init

## 1.3.0 (2019-04-14)
- FIXED: /status/v1/services API access
- FIXED: to ensure JRE 8 is used for puppetdb/puppetserver
- CHANGED: aligned puppetdb/puppetserver system unit files with the official ones

## 1.1.0 (2018-12-09)
- CHANGED: updated for Ubuntu 18.04 Bionic support
- CHANGED: to force system upgrade on new puppet agent initialization
- CHANGED: to use cfhttp service in firewall config
- FIXED: to use "cfssh" instead of "ssh" for repo-related firewall configuration
- NEW: suppot for cfdb::postresql::init_db
- NEW: added cfpuppetserver::client_hosts

## 1.0.1 (2018-04-13)
- CHANGED: disabled automatic deactivation & purging of nodes by default

## 0.12.3 (2018-03-19)
- CHANGED: to load netfilter conntrack modules on bootstrap
- FIXED: to use proper Puppet 5 release deb on bootstrap of clients
- FIXED: to install dirmngr & apt-transport-https package as part of boostrap script

## 0.12.2 (2018-03-15)
- CHANGED: to use cflogsink module for centralized logging, if configured

## [0.12.1](https://github.com/codingfuture/puppet-cfpuppetserver/releases/tag/v0.12.1)
- CHANGED: to use syslog for PuppetServer & PuppetDB
- CHANGED: migrated to Ruby 9K
- NEW: PuppetServer & PuppetDB to version all configuration files for reload detection

## [0.12.0](https://github.com/codingfuture/puppet-cfpuppetserver/releases/tag/v0.12.0)
- CHANGED: migrated to Puppet 5
- CHANGED: to support forced PostgreSQL version

## [0.11.3](https://github.com/codingfuture/puppet-cfpuppetserver/releases/tag/v0.11.3)
- NEW: added cf_purge_node script

## [0.11.2](https://github.com/codingfuture/puppet-cfpuppetserver/releases/tag/v0.11.2)
- FIXED: cfpuppetdb/cfpuppetserver.service TimeoutStartSec to 180sec

## [0.11.1](https://github.com/codingfuture/puppet-cfpuppetserver/releases/tag/v0.11.1)
- FIXED: ExecPostStart -> ExecStartPost in systemd files
- FIXED: to properly wait for cfpuppetserver/cfpuppetdb startup in systemd
- CHANGED: to use Hiera v5 as default
- CHANGED: to require Puppet >= 4.9
- CHANGED: commented out deprecated "environment_data_provider"
- NEW: Puppet 5.x support
- NEW: Ubuntu Zesty support

## [0.11.0]
- Fixed to properly handle PuppetDB-cfdb config
- Added dependency on cfnetwork:firewall anchor for cfpuppetdb/cfpuppetserver
- Enforced public parameter types
- Updated to use cfauth::sudoentry
- Implemented JRuby pool flush on with cf_pupperserver_reload tool
- Added full reload for puppetserver @ systemd
- Added cf_pupperserver_reload for deploy procedure
- Implemented puppetserver & puppetdb restart on package update
- Added --trace to provisioning scripts
- Minor refactoring
- Changed to cache environment until refresh through cf_puppetserver_reload
- Added pupetserver settings_tune
- Added autodiscovery, puppetdb_hosts, puppetserver_hosts and is_arbitrator
    parameters
- Implemented autodiscovery of other instances
- Changed, PuppetServer contacts all PuppetDB instances by default (autodiscovery)
- Added PuppetServer msgpack support
- Updated to new Puppet Function API

## [0.10.5]
- Improved server & client bootstrap scripts
- Fixed minor issues in bootstrap process
- Disabled deprecated PuppetDB configuration directives
- Fixed to properly use PermSize for JVM7 and MetaspaceSize for JVM8+
- Updated minimal memory for PuppetDB & PuppetServer to 512MB
- Updated to cfsystem:0.10.1, cfdb:0.10.1

## [0.10.4]
- Updated CF deps to v0.10.x

## [0.10.3]
- Changed to use `$strict=warning` and `$disable_warnings='deprecations'` by default
- Added control of `$strict` and `$disable_warnings` options
- Fixed validation warnings

## [0.10.2]
- Removed PuppetDB defaults for `node-ttl` and `node-purge-ttl` as it
    led to "vanished" nodes in not so actually maintained deployments.

## [0.10.1]
- Fixed to install activesupport <5.0

## [0.10.0]
- Updated to backward incompatible setup of puppetserver 2.5.0

## [0.9.7]
- Migrated to `cfdb` module for PostgreSQL provisioning and High Availability setup
- Cpmpletely rewritten PuppetDB configuration
- Many parameters change!
- Security enforcement for PuppetDB access authorization

## [0.9.6]
- Disabled scheduled agent runs safety purposes
- Added custom puppetserver.conf to mitigate memory leaks with JRuby tuning

## [0.9.5]
- Updated to Puppet 4.5.0
- Enforced strict mode checking
- Minor fixes
- Added $allow_update_check option
- Fixed minor issues in puppet server bootstrap script
- Updated to latest deps

## [0.9.4]
- Updated cfsystem to 0.9.9
- Changed to install all scripts under /opt/codingfuture/bin
    * cf_r10k_deploy
    * cf_gen_puppet_client_init

## [0.9.3]

- Fixed issues in deploy.sh under some conditions
    - Forcibly added Puppet bin folder to PATH
    - Fixed deploy.sh created by setup_puppetserver.sh bootstrap script

## [0.9.2]

- Fixed use_srv_records and ca_server puppet setting to depend on correct parameters
- Changed to use primary Puppet host for secondary Puppet servers
- Fixed dependency issues when installing Puppet Server from Puppet itself

## [0.9.1]

- Implemented proper 3 level Global Hiera -> Environment Data Provider -> Module Data Provider
  configuration lookup instead of pure Hiera-based
- Moved main PuppetServer to cfsystem module and added support for more paramaters from there
- Added checks for minimal configured RAM of each service
- Added advanced PostgreSQL configuration with SSL support based on Puppet's PKI
- Fixed not to reload PuppetServer on configuration change as it leads to aborted deployment run
- Removed installation of deep_merge gem
- Fixed slave Puppet Server provisioning issues
- Added Puppet environments to etckeeeper ignore
- Fixed to properly disable CA service on slave Puppet Server

lib/puppet/provider/cf_puppetserver/cfprov.rb
#
# Copyright 2016-2019 (c) Andrey Galkin
#


require File.expand_path( '../../../../puppet_x/cf_puppet_server', __FILE__ )


Puppet::Type.type(:cf_puppetserver).provide(
    :cfprov,
    :parent => PuppetX::CfPuppetServer::ProviderBase
) do
    desc "Provider for cfdb_access"
    
    commands :sudo => PuppetX::CfSystem::SUDO
    commands :systemctl => PuppetX::CfSystem::SYSTEMD_CTL
        
    def self.get_config_index
        'cf20puppet1server'
    end

    def self.get_generator_version
        cf_system().makeVersion(__FILE__)
    end
    
    def self.check_exists(params)
        debug("check_exists: #{params}")
        begin
            systemctl(['status', "#{params[:service_name]}.service"])
        rescue => e
            warning(e)
            #warning(e.backtrace)
            false
        end
    end    

    def self.on_config_change(newconf)
        debug('on_config_change')
        
        newconf = newconf[newconf.keys[0]]
        service_name = newconf[:service_name]
        user = 'puppet'
        
        avail_mem = cf_system.getMemory(service_name)
        
        if is_jvm_metaspace
            meta_mem = (avail_mem * 0.2).to_i
            meta_mem = cf_system.fitRange(256, avail_mem, meta_mem)
            meta_param = 'MetaspaceSize'
        else
            meta_mem = (avail_mem * 0.05).to_i
            meta_mem = cf_system.fitRange(256, avail_mem, meta_mem)
            meta_param = 'PermSize'
        end
        
        heap_mem = ((avail_mem - meta_mem) * 0.95).to_i
        
        conf_root_dir = '/etc/puppetlabs/puppetserver'
        conf_dir = "#{conf_root_dir}/conf.d"
        
        need_restart = false
        
        # Service File
        #==================================================
        start_timeout = 180
        stop_timeout = 60

        java_args = [
            "-Xms#{(heap_mem/2).to_i}m",
            "-Xmx#{heap_mem}m",
            "-XX:#{meta_param}=#{(meta_mem/2).to_i}m",
            "-XX:Max#{meta_param}=#{meta_mem}m",
        ]

        default_puppetserver_env = '/etc/default/puppetserver'
        puppet_env = [
            "# This file is auto-generated by cfpuppetserver modules",
            %Q{JAVA_BIN="/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java"},
            %Q{JAVA_ARGS="#{java_args.join(' ')}"},
            %Q{TK_ARGS=""},
            %Q{USER="#{user}"},
            %Q{GROUP="#{user}"},
            %Q{INSTALL_DIR="/opt/puppetlabs/server/apps/puppetserver"},
            %Q{CONFIG="/etc/puppetlabs/puppetserver/conf.d"},
            %Q{BOOTSTRAP_CONFIG="/etc/puppetlabs/puppetserver/services.d/,/opt/puppetlabs/server/apps/puppetserver/config/services.d/"},
            %Q{SERVICE_STOP_RETRIES="#{stop_timeout}"},
            %Q{START_TIMEOUT="#{start_timeout}"},
            %Q{RELOAD_TIMEOUT="#{start_timeout}"},
        ]
        cf_system.atomicWrite(default_puppetserver_env, puppet_env, {:user => user})

        conf_ver = PuppetX::CfSystem.makeVersion([
            conf_root_dir,
            default_puppetserver_env,
            '/etc/puppetlabs/puppet/puppet.conf',
            '/etc/puppetlabs/puppet/puppetdb.conf'
        ])

        content_ini = {
            'Unit' => {
                'Description' => "CF PuppetServer",
            },
            'Service' => {
                '# Package Version' => PuppetX::CfSystem::Util.get_package_version('puppetserver'),
                '# Config Digest' => conf_ver,
                'EnvironmentFile' => default_puppetserver_env,
                'Type' => 'forking',
                'ExecStart' => '/opt/puppetlabs/server/apps/puppetserver/bin/puppetserver start',
                'ExecStop' => '/opt/puppetlabs/server/apps/puppetserver/bin/puppetserver stop',
                'ExecReload' => '/opt/puppetlabs/server/apps/puppetserver/bin/puppetserver reload',
                'ExecStartPost' => "#{PuppetX::CfSystem::WAIT_SOCKET_BIN} 8140 #{start_timeout}",
                'KillMode' => 'process',
                'WorkingDirectory' => conf_root_dir,
                'PIDFile' => '/var/run/puppetlabs/puppetserver/puppetserver.pid',
                'TasksMax' => 4915,
                'TimeoutStartSec' => start_timeout,
                'TimeoutStopSec' => stop_timeout,
                'SuccessExitStatus' => 143,
            },
        }
        
        service_changed = self.cf_system().createService({
            :service_name => service_name,
            :content_ini => content_ini,
            :user => user,
            :cpu_weight => newconf[:cpu_weight],
            :io_weight => newconf[:io_weight],
            :mem_limit => avail_mem,
            :mem_lock => true,
        })
        
        need_restart ||= service_changed
        
        # during migration
        cf_system.maskService('puppetserver')

        #==================================================
        
        if need_restart
            warning(">> reloading #{service_name}")
            systemctl('restart', "#{service_name}.service")
            wait_sock(service_name, 8140)
        end        
    end
end
